federation/executor.go

```diff
@@ -59,72 +94,56 @@ func fetchSchema(ctx context.Context, e ExecutorClient, metadata interface{}) (*
 	})
 }
 
-func NewExecutor(ctx context.Context, executors map[string]ExecutorClient) (*Executor, error) {
-	return NewExecutorWithOptionalArgs(ctx, executors, nil)
+type CustomExecutorArgs struct {
+	SchemaSyncer             SchemaSyncer
+	OptionalArgs             interface{}
 }
 
-func NewExecutorWithOptionalArgs(ctx context.Context, executors map[string]ExecutorClient, optionalArgs interface{}) (*Executor, error) {
-	// Fetches the schemas from the executors clients
-	schemas := make(map[string]*introspectionQueryResult)
-	for server, client := range executors {
-		resp, err := fetchSchema(ctx, client, optionalArgs)
-		schema := resp.Result
-		if err != nil {
-			return nil, oops.Wrapf(err, "fetching schema %s", server)
-		}
-
-		var iq introspectionQueryResult
-		if err := json.Unmarshal(schema, &iq); err != nil {
-			return nil, oops.Wrapf(err, "unmarshaling schema %s", server)
-		}
-
-		schemas[server] = &iq
+func NewExecutor(ctx context.Context, executors map[string]ExecutorClient, c *CustomExecutorArgs) (*Executor, error) {
+	if c.SchemaSyncer == nil {
+		c.SchemaSyncer = newIntrospectionSchemaSyncer(ctx, executors, c.OptionalArgs)
 	}
 
-	types, err := convertSchema(schemas)
+	planner, err := c.SchemaSyncer.FetchPlanner(ctx, c.OptionalArgs)
 	if err != nil {
-		return nil, oops.Wrapf(err, "converting schemas error")
+		return nil, oops.Wrapf(err, "failed to load schema")
 	}
-
-	introspectionSchema := introspection.BareIntrospectionSchema(types.Schema)
-	introspectionServer := &Server{schema: introspectionSchema}
-
-	executors["introspection"] = &DirectExecutorClient{Client: introspectionServer}
-	schema, err := introspection.RunIntrospectionQuery(introspection.BareIntrospectionSchema(introspectionServer.schema))
-
-	var iq introspectionQueryResult
-	if err := json.Unmarshal(schema, &iq); err != nil {
-		return nil, oops.Wrapf(err, "unmarshaling introspection schema")
-	}
-
-	schemas["introspection"] = &iq
-	types, err = convertSchema(schemas)
+	interval, err := c.SchemaSyncer.GetRefetchInterval(ctx)
```

Comment:
This feels a little weird to me since we have the GetInterval inside the syncer while the poller is outside of the syncer.
I am okay with either to put all the interval and poller stuff to the syncer or the other way.